dpcpp/base/dim3.dp.hpp

```diff
@@ -0,0 +1,96 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_DPCPP_BASE_DIM3_DP_HPP_
+#define GKO_DPCPP_BASE_DIM3_DP_HPP_
+
+
+#include <CL/sycl.hpp>
+
+
+namespace gko {
+namespace kernels {
+namespace dpcpp {
+
+
+/**
+ * dim3 is a cuda-like dim3 for sycl-range, which provides the same ordering as
+ * cuda and gets the sycl-range in reverse ordering.
+ */
+struct dim3 {
```

Comment:
We are building a lot of complexity here (also the config selection dim3 integration), so I want to ask: Why do we need this additional wrapper? Can't we use SYCL primitives directly?

Reply:
Yes, we can use the SYCL native `sycl::range<3>`
but it will gives different view on kernel launch and build the sycl::range
for example, we use `kernel<<<32, 32>>>` and sycl will need to use
`kernel(sycl::nd_range<3>(sycl::range<3>(1, 1, 32) * sycl::range<3>(1, 1, 32), sycl::range<3>(1, 1, 32)))`
but with dim3, we can still use `kernel(32, 32)`
and in the beginning, I would like to reduce the difference from cuda to dpcpp

Reply:
In which places does that actually matter? Most of the time, we are using one-dimensional kernels, except for SpMV (multiple columns), some Dense kernels (2D) and Jacobi. Wouldn't you use `sycl::ndrange<1>(...)` directly?

Reply:
I think the mapping will be different. At least, it will require us to develop own dpct to avoid any conversion like threadIdx -> get_local_id(3)

Reply:
and it will gives inconsistent index sense between the kernel1D/2D/3D.
`range<1>(x)` `get_local_id(0)` is x and x is contiguous as cuda
`range<2>(x, y)` `get_local_id(0)` is still x but x is not contiguous
or `range<2>(y, x)` x is contiguous but x needs `get_local_id(1)`

Reply:
But that only matters on the launch side, right? I would assume that the reason for this is that it maps more cleanly to the nested for-loop model
```
for (; i < range[0]; i++)
    for (; j < range[1]; j++)
        for (; k < range[2]; k++)
```

Reply:
it's not only affect launch side but the kernel index for threads